jerry-debugger/src/lib/protocol-handler.ts

```diff
@@ -0,0 +1,251 @@
+      this.abort('compressed pointer must be 2 or 4 bytes long');
+    }
+
+    if (this.version !== JERRY_DEBUGGER_VERSION) {
+      this.abort('incorrect target debugger version detected: ' + this.version
+                 + ' expected: ' + JERRY_DEBUGGER_VERSION);
+    }
+  }
+
+  onByteCodeCP(data: Uint8Array) {
+    console.log('[Byte Code CP]');
+    const frame = this.stack.pop();
+    if (!frame) {
+      throw new Error('missing parser stack frame');
+    }
+
+    const byteCodeCP = this.decodeMessage('C', data, 1)[0];
+    const func = new ParsedFunction(byteCodeCP, frame);
+
+    this.functions[byteCodeCP] = func;
+    if (this.stack.length > 0) {
+      return;
+    }
+
+    // FIXME: it seems like this is probably unnecessarily keeping the
+    //   whole file's source to this point?
+    func.source = this.source.split(/\r\n[\r\n]/);
+    func.sourceName = this.sourceName;
+  }
+
+  onParseFunction(data: Uint8Array) {
+    console.log('[Parse Function]');
+    const position = this.decodeMessage('II', data, 1);
+    this.stack.push({
+      isFunc: true,
+      line: position[0],
+      column: position[1],
+      name: cesu8ToString(this.functionName),
+      source: this.source,
+      sourceName: this.sourceName,
+      lines: [],
+      offsets: [],
+    });
+    this.functionName = undefined;
+    return;
+  }
+
+  onBreakpointList(data: Uint8Array) {
+    console.log('[Breakpoint List]');
+    if (data.byteLength % 4 !== 1 || data.byteLength < 1 + 4) {
+      throw new Error('unexpected breakpoint list message length');
+    }
+
+    let array: Array<number>;
+    if (data[0] === SP.JERRY_DEBUGGER_BREAKPOINT_LIST) {
+      array = this.stack[this.stack.length - 1].lines;
+    } else {
+      array = this.stack[this.stack.length - 1].offsets;
+    }
+
+    for (let i = 1; i < data.byteLength; i += 4) {
+      array.push(this.decodeMessage('I', data, i)[0]);
+    }
+    return;
+  }
+
+  onSourceCode(data: Uint8Array) {
+    console.log('[Source Code]');
+    this.sourceData = concatUint8Arrays(this.sourceData, data);
+    if (data[0] === SP.JERRY_DEBUGGER_SOURCE_CODE_END) {
+      this.source = cesu8ToString(this.sourceData);
+      if (this.delegate.onScriptParsed) {
+        this.delegate.onScriptParsed({
+          'scriptId': ++this.lastScriptID,
+          'name': this.sourceName || '',
+          'endLine': this.source.split(/\r\n[\r\n]/).length,
+          'endColumn': 0,
+          'hasSourceURL': false,
+        });
+      }
+    }
+  }
+
+  onSourceCodeName(data: Uint8Array) {
+    console.log('[Source Code Name]');
+    this.sourceNameData = concatUint8Arrays(this.sourceNameData, data);
+    if (data[0] === SP.JERRY_DEBUGGER_SOURCE_CODE_NAME_END) {
+      this.sourceName = cesu8ToString(this.sourceNameData);
+      // assumption: this will be completed before source and included in the
+      //   onScriptParsed delegate function called in onSourceCode
+    }
+  }
+
+  parseData(data: ArrayBuffer) {
```

Comment:
Why take `ArrayBuffer` and not `Uint8Array` here?
I would push the conversion to `Uint8Array` out one layer.
`ArrayBuffer` is a bit cumbersome to work with because you always have to create a view before you can do anything with it.